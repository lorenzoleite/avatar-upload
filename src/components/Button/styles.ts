import styled, { css } from 'styled-components'

export const StyledButton = styled.button`
  ${({ theme }) => css`
    border: none;
    padding: 0.2rem 3.6rem;
    border-radius: 1.6rem;
    background-color: ${theme.colors['gray.600']};
    color: ${theme.colors.white};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.weights.medium};
    line-height: ${theme.lineHeights.small};
  `}
`
