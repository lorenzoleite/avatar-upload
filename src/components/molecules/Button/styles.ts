import styled, { DefaultTheme, css } from 'styled-components';
import { ButtonProps } from '.';

const modifiers = {
  underline: (theme: DefaultTheme) => css`
    font-size: ${theme.font.size.medium};
    line-height: ${theme.font.lineHeight.medium};
    text-decoration: underline;

    padding: 0;
    border-radius: 0;

    background-color: transparent;
    color: ${theme.color['gray.600']};

    &:hover {
      background-color: transparent;
      color: ${theme.color['gray.400']};
    }
  `,
  icon: (theme: DefaultTheme) => css`
    font-family: none;
    font-size: 0;
    line-height: 0;
    text-decoration: none;

    padding: 0.2rem;
    border-radius: 0;

    clip-path: circle();
    background-color: transparent;

    &:hover {
      padding: 0.2rem;
      background-color: ${theme.color['gray.200']};
    }
  `,
};

export const StyledButton = styled.button<ButtonProps>`
  ${({ theme, variant }) => css`
    cursor: pointer;
    font-family: ${theme.font.family};
    font-size: ${theme.font.size.small};
    font-weight: ${theme.font.weight.medium};
    line-height: ${theme.font.lineHeight.small};

    padding: 0.2rem 3.6rem;
    border: none;
    border-radius: 1.6rem;

    background-color: ${theme.color['gray.600']};
    color: ${theme.color.white};

    transition: 0.2s ease-in-out;

    &:hover {
      background-color: ${theme.color['gray.400']};
    }

    ${variant === 'underline' && modifiers.underline(theme)}
    ${variant === 'icon' && modifiers.icon(theme)}
  `}
`;
