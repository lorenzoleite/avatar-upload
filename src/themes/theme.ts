export const theme = {
  colors: {
    white: '#FFFFFF',

    'gray.600': '#3D485F',
    'gray.500': '#495567',
    'gray.400': '#677489',
    'gray.300': '#C7CDD3',
    'gray.200': '#C3CBD5',
    'gray.100': '#F2F5F8',

    'blue.200': '#3F80FF',
    'blue.100': '#B9D1FF',

    'orange.100': '#C64D32',
  },
  font: {
    family: 'Inter, sans-serif',
    weights: {
      normal: 400,
      medium: 500,
    },
    sizes: {
      small: '1.4rem',
      medium: '1.6rem',
    },
  },
  borders: {
    small: '0.8rem',
    medium: '1.6rem',
  },
  lineHeights: {
    small: '2.52rem',
    medium: '2.88rem',
  },
};
