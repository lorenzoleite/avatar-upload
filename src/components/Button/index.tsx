import React, { ButtonHTMLAttributes } from 'react';

import { StyledButton } from './styles';

export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  variant?: 'solid' | 'underline' | 'icon';
};

export function Button({ children, ...props }: ButtonProps) {
  return <StyledButton {...props}>{children}</StyledButton>;
}
