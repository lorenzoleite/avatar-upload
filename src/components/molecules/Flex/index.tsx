import React from 'react';

import { StyledFlex } from './styles';

export type FlexProps = {
  children: React.ReactNode;
  direction?: 'row' | 'column';
  justify?: 'center' | 'flex-start' | 'flex-end' | 'space-between' | 'space-around' | 'space-evenly';
  align?: 'center' | 'flex-start' | 'flex-end' | 'baseline' | 'stretch';
  gap?: string;
  width?: string;
};

export function Flex({ children, ...props }: FlexProps) {
  return <StyledFlex {...props}>{children}</StyledFlex>;
}
