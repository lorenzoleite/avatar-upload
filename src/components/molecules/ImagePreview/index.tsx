import React from 'react';

import { StyledImagePreview } from './styles';

export type ImagePreviewProps = {
  children?: React.ReactNode;
  marginRight?: string;
};

export function ImagePreview({ children, ...props }: ImagePreviewProps) {
  return <StyledImagePreview {...props}>{children}</StyledImagePreview>;
}
