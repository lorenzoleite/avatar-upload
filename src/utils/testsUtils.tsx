import React from 'react';
import fetch from 'node-fetch';

import { ThemeProvider } from 'styled-components';
import { render, RenderResult } from '@testing-library/react';

import { theme } from '../themes/theme';

export const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);

export const imageFile = await fetch(
  'https://images.unsplash.com/photo-1661956602868-6ae368943878?ixlib=rb-4.0.3&ixid=MnwxMjA3fDF8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=234&q=16',
)
  .then(response => response.blob())
  .then(blob => new File([blob], 'image.jpg', { type: 'image/jpeg' }));
