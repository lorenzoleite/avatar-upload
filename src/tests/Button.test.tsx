import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import { describe, it } from 'vitest';

import { Button } from '../components/molecules/Button';
import { renderWithTheme } from '../utils/testsUtils';
import { theme } from '../themes/theme';

describe('Button', () => {
  it('should render the button', () => {
    renderWithTheme(<Button>Save</Button>);

    expect(screen.getByRole('button')).toBeInTheDocument();
  });
  it('should render the button with solid variant', () => {
    renderWithTheme(<Button variant="solid">Save</Button>);

    expect(screen.getByRole('button')).toBeInTheDocument().toHaveStyle({
      'background-color': theme.color['gray.600'],
    });
  });
  it('should render the button with underline variant', () => {
    renderWithTheme(<Button variant="underline">Save</Button>);

    expect(screen.getByRole('button')).toBeInTheDocument().toHaveStyle({
      'background-color': 'transparent',
      'text-decoration': 'underline',
    });
  });
  it('should render the button with icon variant', () => {
    renderWithTheme(<Button variant="icon">Save</Button>);

    expect(screen.getByRole('button')).toBeInTheDocument().toHaveStyle({
      'background-color': 'transparent',
      'text-decoration': 'none',
      'clip-path': 'circle()',
    });
  });
});
