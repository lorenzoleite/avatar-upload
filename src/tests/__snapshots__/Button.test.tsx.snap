// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should render the button 1`] = `
<button
  class="sc-beqWaB izZfKp"
>
  Save
</button>
`;

exports[`Button > should render the button 2`] = `
<button
  class="sc-beqWaB izZfKp"
>
  Save
</button>
`;

exports[`Button > should render the button with icon variant 1`] = `
<button
  class="sc-beqWaB itQUyh"
>
  Save
</button>
`;

exports[`Button > should render the button with solid variant 1`] = `
<button
  class="sc-beqWaB izZfKp"
>
  Save
</button>
`;

exports[`Button > should render the button with underline variant 1`] = `
<button
  class="sc-beqWaB gbwlyn"
>
  Save
</button>
`;

exports[`Button > should render the button with variant 1`] = `
<button
  class="sc-beqWaB izZfKp"
>
  Save
</button>
`;
